src/openrct2/rct1/S4Importer.cpp

```diff
@@ -1955,16 +1955,222 @@ class S4Importer final : public IParkImporter
+                dst2->SetPhotoTimeout(src2->GetPhotoTimeout());
+                // Skipping IsHighlighted()
+
+                // TODO: Import Door A and Door B states.
+
+                break;
+            }
+            case TILE_ELEMENT_TYPE_SMALL_SCENERY:
+            {
+                auto dst2 = dst->AsSmallScenery();
+                auto src2 = src->AsSmallScenery();
+
+                uint8_t entryIndex = _smallSceneryTypeToEntryMap[src2->GetEntryIndex()];
+                dst2->SetEntryIndex(entryIndex);
+                dst2->SetAge(src2->GetAge());
+                dst2->SetSceneryQuadrant(src2->GetSceneryQuadrant());
+                dst2->SetPrimaryColour(RCT1::GetColour(src2->GetPrimaryColour()));
+                if (src2->NeedsSupports())
+                    dst2->SetNeedsSupports();
+
+                // Copied from [rct2: 0x006A2956]
+                switch (src2->GetEntryIndex())
+                {
+                    case RCT1_SCENERY_GEOMETRIC_SCULPTURE_1:
+                    case RCT1_SCENERY_GEOMETRIC_SCULPTURE_2:
+                    case RCT1_SCENERY_GEOMETRIC_SCULPTURE_3:
+                    case RCT1_SCENERY_GEOMETRIC_SCULPTURE_4:
+                    case RCT1_SCENERY_GEOMETRIC_SCULPTURE_5:
+                        dst2->SetSecondaryColour(COLOUR_WHITE);
+                        break;
+                    case RCT1_SCENERY_TULIPS_1:
+                    case RCT1_SCENERY_TULIPS_2:
+                        dst2->SetPrimaryColour(COLOUR_BRIGHT_RED);
+                        dst2->SetSecondaryColour(COLOUR_YELLOW);
+                }
+
+                break;
+            }
+            case TILE_ELEMENT_TYPE_ENTRANCE:
+            {
+                auto dst2 = dst->AsEntrance();
+                auto src2 = src->AsEntrance();
+
+                dst2->SetEntranceType(src2->GetEntranceType());
+                dst2->SetRideIndex(src2->GetRideIndex());
+                dst2->SetStationIndex(src2->GetStationIndex());
+                dst2->SetSequenceIndex(src2->GetSequenceIndex());
+
+                if (src2->GetEntranceType() == ENTRANCE_TYPE_PARK_ENTRANCE)
+                {
+                    uint8_t pathType = src2->GetPathType();
+                    if (pathType == 0)
+                    {
+                        pathType = RCT1_FOOTPATH_TYPE_TARMAC_GRAY;
+                    }
+                    uint8_t entryIndex = _pathTypeToEntryMap[pathType];
+                    dst2->SetPathType(entryIndex & 0x7F);
+                }
+
+                break;
+            }
+            case TILE_ELEMENT_TYPE_WALL:
+            {
+                auto dst2 = dst->AsWall();
+                auto src2 = src->AsWall();
+
+                dst2->SetEntryIndex(src2->GetEntryIndex());
+                dst2->SetSlope(src2->GetSlope());
+                dst2->SetPrimaryColour(RCT1::GetColour(src2->GetRCT1WallColour()));
+                dst2->SetRawRCT1Data(src2->GetRawRCT1WallTypeData());
```

Comment:
This should be split into stuff like SetColour, SetAnimation.

Reply:
No, it shouldn't. Because it hasn't got anything to do with colours or animation.

What RCT1 saved in vars 5, 6 and 7 (which we call colour_1, animation and colour_3) is the entry index for the wall type of every edge, using 6 bits for each. In other words, RCT1 used only one WallElement per tile. The information is copied, then FixWalls extracts the information again, removes the temporary WallElement and inserts 1-4 WallElements of the correct entry index. (You even wrote that function.)

Reply:
Call it what it actually means then, the new tile element struct shouldn't have a method related to RCT1 on it.

Reply:
Sigh. It HAS to temporarily store RCT1 data, so an RCT1-named method is appropriate. I just explained why. Did you really take a proper look at the code before you said that?

I'd love not to need this, but it would require a far more complex set-up - one _I_ cannot properly program. As it stands, the code can only import one RCT12TileElement to one TileElement, not one-to-four.